event_listener/dfusion_db/event_receivers.py

```diff
@@ -0,0 +1,80 @@
+from abc import abstractmethod
+
+from django_eth_events.chainevents import AbstractEventReceiver
+from .event_pusher import post_deposit, post_transition
+
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class GenericEventReceiver(AbstractEventReceiver):
+    name = None
+
+    def ensure_name(self, _name):
+        return self.name == _name
+
+    def parse_event(self, decoded_event):
+        res = {param['name']: param['value'] for param in decoded_event['params']}
+        logging.info("{} received {}".format(self.name, res))
+        return res
+
+    def save(self, decoded_event, block_info=None):
+        if not self.ensure_name(decoded_event['name']):
+            return
+
+        parsed_event = self.parse_event(decoded_event)
+
+        self.real_save(parsed_event, block_info)
+
+    def rollback(self, decoded_event, block_info=None):
+        if not self.ensure_name(decoded_event['name']):
+            return
+
+        self.real_rollback(decoded_event, block_info)
+
+    @abstractmethod
+    def real_save(self, decoded_event, block_info=None):
+        pass
+
+    @abstractmethod
+    def real_rollback(self, decoded_event, block_info=None):
+        pass
+
+
+class DepositReceiver(GenericEventReceiver):
```

Comment:
In general, I prefer using composition over inheritance unless there is a good reason (generally composition gives more design flexibility and ends up being simpler as it's not trying to shoehorn different things into a common base). Happy to discuss this principle more in detail.

Here, I'd suggest to give GenericEventReceiver access to a mapping of String => AbstractEventReceiver and have an abstract event receiver for each type of event (DepositReceiver, StateTransitionReceiver).

This would avoid creating these "real_save", "real_deposit methods".

The GenericEventReceiver could be called something like DfusionEventDispatcher or something along those lines.

Reply:
I considered using the old dictionary trick to avoid a long ugly if-elif-...-else and had decided against it assuming you wouldn’t like it. There is something that one need to be aware of when using dictionaries like this with classes as values. The python interpreter evaluates all of them when the mapping is declared. This means if we try to stuff a Deposit and StateTransition event into this, we could get failure when the event parameters aren’t correct.

Reply:
I’m any case I am in favour of the dispatcher idea, but may need to be careful about using classes as values.

Reply:
I’m any case I am in favour of the dispatcher idea, but may need to be careful about using classes as values.

Reply:
I’m any case I am in favour of the dispatcher idea, but may need to be careful about using classes as values.

Reply:
We also need to consider what our dependencies allow us to do.